src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10755,97 +10844,127 @@ private ExpressionSyntax ParseRegularStackAllocExpression()
-                PeekToken(1).Kind != SyntaxKind.EqualsGreaterThanToken)
+            var modifiers = _pool.Allocate();
+
+            while (true)
             {
-                asyncToken = this.EatContextualToken(SyntaxKind.AsyncKeyword);
-                asyncToken = CheckFeatureAvailability(asyncToken, MessageID.IDS_FeatureAsync);
-                IsInAsync = true;
-            }
+                if (this.CurrentToken.Kind == SyntaxKind.StaticKeyword)
+                {
+                    var staticKeyword = this.EatToken(SyntaxKind.StaticKeyword);
+                    staticKeyword = CheckFeatureAvailability(staticKeyword, MessageID.IDS_FeatureStaticAnonymousFunction);
+                    modifiers.Add(staticKeyword);
+                    continue;
+                }
 
-            var result = ParseLambdaExpression(asyncToken);
+                if (this.CurrentToken.ContextualKind == SyntaxKind.AsyncKeyword &&
+                    this.PeekToken(1).Kind != SyntaxKind.EqualsGreaterThanToken)
+                {
+                    var asyncToken = this.EatContextualToken(SyntaxKind.AsyncKeyword);
+                    asyncToken = CheckFeatureAvailability(asyncToken, MessageID.IDS_FeatureAsync);
+                    modifiers.Add(asyncToken);
+                    continue;
+                }
 
-            IsInAsync = parentScopeIsInAsync;
+                break;
+            }
+
+            var result = modifiers.ToList();
+            _pool.Free(modifiers);
             return result;
         }
 
-        private LambdaExpressionSyntax ParseLambdaExpression(SyntaxToken asyncToken)
+        private LambdaExpressionSyntax ParseLambdaExpression()
         {
-            if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
-            {
-                var paramList = this.ParseLambdaParameterList();
-                var arrow = this.EatToken(SyntaxKind.EqualsGreaterThanToken);
-                arrow = CheckFeatureAvailability(arrow, MessageID.IDS_FeatureLambda);
-                var (block, expression) = ParseLambdaBody();
+            var parentScopeIsInAsync = this.IsInAsync;
+            var result = parseLambdaExpressionWorker();
+            this.IsInAsync = parentScopeIsInAsync;
+            return result;
 
-                return _syntaxFactory.ParenthesizedLambdaExpression(
-                    asyncToken, paramList, arrow, block, expression);
-            }
-            else
+            LambdaExpressionSyntax parseLambdaExpressionWorker()
             {
-                var name = this.ParseIdentifierToken();
-                var arrow = this.EatToken(SyntaxKind.EqualsGreaterThanToken);
-                arrow = CheckFeatureAvailability(arrow, MessageID.IDS_FeatureLambda);
+                var modifiers = ParseAnonymousFunctionModifiers();
+                if (modifiers.Any((int)SyntaxKind.AsyncKeyword))
+                {
+                    this.IsInAsync = true;
+                }
```

Comment:
`IsInAsync = modifiers.Any((int)SyntaxKind.AsyncKeyword);`

Reply:
See above.  very very intentional.  We do not want to set it back to 'false' in an inner lambda.